### ISSUE TITLE
Replaced usage of dicts with nonsense keys with sets in epoll

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6685]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Alex reported|
|-|-|
|Trac ID|trac#6685|
|Type|enhancement|
|Created|2013-08-19 02:20:04Z|
|Branch|https://github.com/twisted/twisted/tree/epoll-sets-6685|

This should save memory, and with it a bit of CPU time.

Also replaced some usage of bound methods as attributes with normal methods, this is faster on PyPy and allows for including a docstring.

Attachments:

* [epoll-sets.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0e8/0e8fac450e57fd88bae4eea99d5326d6b6011087/964820bd48aa77942e980bc72371c4c89489b634.diff) (3866 bytes) - added by Alex on 2013-08-19 02:20:13Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6685 6685
type__enhancement enhancement
reporter__Alex Alex
priority__normal normal
milestone__ 
branch__branches_epoll_sets_6685 branches/epoll-sets-6685
branch_author__lvh lvh
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1376878804000000 1376878804000000
changetime__1380457169000000 1380457169000000
version__None None
owner__lvh lvh

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|Alex commented|
|-|-|

A benchmark demonstrating the claim that `sets` will be faster than `dicts`:
```
Alexanders-MacBook-Pro:twisted alex_gaynor$ pypy -mtimeit -s "s = {2: 1}" "s[1] = 1; del s[1]"
100000000 loops, best of 3: 0.0196 usec per loop
Alexanders-MacBook-Pro:twisted alex_gaynor$ pypy -mtimeit -s "s = {2: 1}" "s[1] = 1; del s[1]"
100000000 loops, best of 3: 0.0196 usec per loop
Alexanders-MacBook-Pro:twisted alex_gaynor$ pypy -mtimeit -s "s = {2: 1}" "s[1] = 1; del s[1]"
100000000 loops, best of 3: 0.0196 usec per loop
Alexanders-MacBook-Pro:twisted alex_gaynor$
Alexanders-MacBook-Pro:twisted alex_gaynor$
Alexanders-MacBook-Pro:twisted alex_gaynor$ pypy -mtimeit -s "s = set([2])" "s.add(1); s.remove(1)"
100000000 loops, best of 3: 0.0182 usec per loop
Alexanders-MacBook-Pro:twisted alex_gaynor$ pypy -mtimeit -s "s = set([2])" "s.add(1); s.remove(1)"
100000000 loops, best of 3: 0.0182 usec per loop
Alexanders-MacBook-Pro:twisted alex_gaynor$ pypy -mtimeit -s "s = set([2])" "s.add(1); s.remove(1)"
100000000 loops, best of 3: 0.0183 usec per loop
```

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)|@lvh set owner to @lvh|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_3"></a>|@lvh commented|
|-|-|

(In [39924]) Branching to 'epoll-sets-6685'

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_4"></a>|@lvh commented|
|-|-|

This doesn't affect the unit tests, but twistedchecker does complain because the new methods don't document their argument or return value:

```
************* Module twisted.internet.epollreactor
W9202:160,4:_ContinuousPolling.isReading: Missing epytext markup @param for argument "fd"
W9204:160,4:_ContinuousPolling.isReading: Missing epytext markup @return for return value
W9202:168,4:_ContinuousPolling.isWriting: Missing epytext markup @param for argument "fd"
W9204:168,4:_ContinuousPolling.isWriting: Missing epytext markup @return for return value
```

(http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/epoll-sets-6685)

Since these changes are quite small, I'll write them myself.

Thank you for your contribution!

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_5"></a>|@lvh removed owner|
|-|-|

Resubmitting for review. (Please use the branch, not the patch.)

There's a bunch of places epollreactor uses C{FileDescriptor}. I think it should use L{FileDescriptor <twisted.internet.abstract.FileDescriptor>} instead. I don't know if that should go on a different ticket or not.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_6"></a>|@lvh commented|
|-|-|

I've re-ran the buildbots:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/epoll-sets-6685

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|Alex commented|
|-|-|

Not sure if I'm allowed to review this, but if I am: looks good to me.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_8"></a>|@lvh set owner to @lvh|
|-|-|

I think that's okay, since we reviewed each others changes, not our own. I'll go ahead and merge when I get a minute.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_9"></a>|@lvh set status to `closed`|
|-|-|

(In [40137]) 
Merge epoll-sets-6685: use sets in epoll reactor

Author: Alex, lvh
Reviewer: lvh, Alex
Fixes: [#6685](https://github.com/twisted/twisted/issues/6685)

The epoll reactor previously used dicts when it semantically meant sets. This was presumably done because sets haven't always existed in Python. Since we no longer support versions of Python where sets are not available, we just use sets all the time now.

Apparently this is also slightly faster in PyPy. Yay!
